QEfficient/base/modeling_qeff.py

```diff
@@ -5,96 +5,284 @@
+        for key, value in compiler_options.items():
+            option = "-" + key.replace("_", "-")
+            if isinstance(value, bool):
+                if value:
+                    command.append(option)
+                continue
+            command.append(f"{option}={value}")
+        compile_hash = hashlib.sha256(to_hashable(command))
+
+        # Write specializations.json file
+        if specializations is not None:
+            specializations_json = compile_dir / "specializations.json"
+            with open(specializations_json, "w") as fp:
+                json.dump(
+                    {"specializations": [{k: str(v) for k, v in spec.items()} for spec in specializations]},
+                    fp,
+                    indent=4,
+                )
+            command.append(f"-network-specialization-config={specializations_json}")
+            compile_hash.update(to_hashable(specializations))
+
+        # Write custom_io.yaml file
+        if custom_io:
+            custom_io_yaml = compile_dir / "custom_io.yaml"
+            with open(custom_io_yaml, "w") as fp:
+                for io_name, dtype in custom_io.items():
+                    fp.write(f" - IOName: {io_name}\n   Precision: {dtype}\n\n")
+            command.append(f"-custom-IO-list-file={custom_io_yaml}")
+            compile_hash.update(to_hashable(custom_io))
+
+        # Write mdp_config.json file
+        if mdp_ts_num_devices > 1:
+            num_cores = compiler_options.get("aic_num_cores", 16)
+            mdp_ts_json = compile_dir / f"mdp_ts_{mdp_ts_num_devices}.json"
+            with open(mdp_ts_json, "w") as fp:
+                json.dump(
+                    {
+                        "connections": [{"devices": list(range(mdp_ts_num_devices)), "type": "p2p"}],
+                        "partitions": [
+                            {
+                                "name": "Partition0",
+                                "devices": [{"deviceId": d, "numCores": num_cores} for d in range(mdp_ts_num_devices)],
+                            }
+                        ],
+                    },
+                    fp,
+                    indent=4,
+                )
+            command.append(f"-mdp-load-partition-config={mdp_ts_json}")
+            compile_hash.update(to_hashable({"mdp_ts_num_devices": mdp_ts_num_devices}))
+
+        # Check if already compiled
+        compile_hash = compile_hash.hexdigest()[:16]
+        qpc_path = qpc_path.with_name(qpc_path.name + "-" + compile_hash)
+        if qpc_path.is_dir():
+            if (qpc_path / "programqpc.bin").is_file():
+                self.qpc_path = qpc_path
+                return qpc_path
+            # Probably compilation failure last time, delete directory to start over
+            shutil.rmtree(qpc_path)
+
+        command.append(f"-aic-binary-dir={qpc_path}")
+        logger.info(f"Running compiler: {' '.join(command)}")
+        subprocess.run(command).check_returncode()
```

Comment:
Shouldnt we add a trycatch for subprocess.run? if it fails qeff would fail with an uncaught exception right?

Reply:
We don't have to try/catch, because even after catching, what can we do? We could only re-raise as a different error. Better to propagate the same error to the user.

Reply:
I think it would be better if we can catch and wrap the error around the stage at which the error is observed. Otherwise, user would need to figure out from where the error is coming from.

Reply:
By wrap around, if you mean add some descriptive message to the user that the compiler failed, we can do that. I'll be adding that.